### ISSUE TITLE
Copyright?

### ISSUE THREAD

**Issue by jfranknichols** [CineDev/ParagraphTextKit#2]
I see you have included a copyright with all rights reserved. Is this your intent, or are you planning to open this with an MIT or other open source copyright? I would like to include this in my project (MIT?) but I will not be able to if the copyright issue exists.

**Comment by CineDev** [comment]
No, I have no intentions to prevent anyone from using this code. Haven't think of copyright at all, that's the default Xcode template ) Just tell me how can I help you to include this code into your project.

**Comment by jfranknichols** [comment]
If you want to share it, I recommend a Creative Common Copyright. This site ill ask you a couple questions, and provide you with a wording that is appropriate. You can the just copy and paste it into your code in place of the default. 

Thank you for sharing.

https://creativecommons.org/share-your-work/ <https://creativecommons.org/share-your-work/>

**Comment by CineDev** [comment]
What about MIT license? Updated the project to 1.2.3 version

**Comment by jfranknichols** [comment]
MIT is fine, basically the same. Thank you.

**Comment by jfranknichols** [comment]
Sorry to bother you again. 

Is it possible to programmatically update a paragraph? 

By this I mean, if the user does an undo on an edit/insert/or remove of a paragraph can I pass that change back, or do I need to reload the presentedParagraphs?

The obvious reason I am interested is I expect the presentedParagraphs to be as much as several MB and I don’t want to move the much data every edit. I am assuming I have to pass in the AttributedString by value and I can’t pass it in by reference.

And if I modify the presentedParagraphs will your kit recognize the change, or is that only read at initialization?

Thanks,

Frank

**Comment by CineDev** [comment]
For undo/redo support you don't need to store the initial state, since the undo action will just edit the NSTextStorage object and you will get the ParagraphDelegate notification like during the regular editing.
If for any reason you still need to store the initial state and to recover it during the undo action, I'd suggest you to store the state wherever you want and when the paragraph delegate will fire the 'didChangeParagraphs' method, check if the undo manager is undoing and do what you need (if undoManager?.isUndoing == true).
But I don't recommend this approach because thus creates another source of truth. Even if it looks like this source of truth is synced with the text storage, technically it doesn't and this might lead to unexpected behaviour in some circumstances. And it'll be really hard to debug.

**Comment by jfranknichols** [comment]
Sorry to keep bothering you.

I am “new” to Swift and MacOS. 

The paragraph data is coming from CoreData. I was working on the assumption that my CoreData model is my source of truth, and I was to keep it up to date using the debate methods (insertedParagraph.removedParagraph and editedParagraph) as changes are occurring) obviously not ever character, but at least periodically. My application is an editor for writing novels, etc. If I wait too long and a crash or power outage occurs the user will not be happy.

Can I assume the presentedParagraphs are the source of truth, or are you referring to the TextStorage? 

Thank you for your patience and help.

**Comment by CineDev** [comment]
presentedParagraphs is the source of truth BEFORE you put them into the ParagraphTextStorage. It's the initial state of your model. In your case — it's the representation of your CoreData model. But since the ParagraphTextStorage has started editing its own model, it becomes the only source of truth for your CoreData model. Look at the ParagraphTextStorageDelegate protocol definition: presentedParagraphs has 'get' attribute, not having 'set', so it's not possible for a ParagraphTextStorage object to edit your model. It relies on ParagraphTextStorageDelegate's didChangeParagraphs method where you should update your model according to the changes been made inside the ParagraphTextStorage.
In other words, presentedParagraphs are used only to create the initial state of the ParagraphTextStorage (say, to display paragraphs stored in CoreData, to display the text stored in a file), not to mutate the initial model. You edit the Core Data model in didChangeParagraphs method by yourself. ParagraphTextStorage is only notifies you of the changes been made in it.

**Comment by jfranknichols** [comment]
Thank you, that answered my questions. I appreciate your time.

**Comment by CineDev** [comment]
You're welcome!
